File: server-refactor/src/builds/dto/create-build.dto.ts
import { Type } from "class-transformer";
import { IsArray, IsBoolean, IsString, Length, ValidateNested } from "class-validator";
import { CreateBuildComponentDto } from "./create-build-component.dto";

export class CreateBuildDto {
    @IsString()
    @Length(1, 100)
    name: string;

    @IsString()
    description: string;

    @IsBoolean()
    isPublic: boolean;

    @IsArray()
    @ValidateNested()
    @Type(() => CreateBuildComponentDto)
    components: CreateBuildComponentDto[];
}

File: client/app/src/main/java/com/derlados/computer_configurator/services/builds/dto/CreateBuildDto.kt
package com.derlados.computer_configurator.services.builds.dto

import com.derlados.computer_configurator.entities.build.Build

/**
 * Данные о сборке для передачи в запросе, так как большинство данные является лищними
 * name - имя сборки
 * desc - описание сборки
 * components - массив пар <id комлпектующего, количество комплектующих>
 */
class CreateBuildDto(build: Build) {
    val name: String
    val description: String
    val isPublic: Boolean
    val components: ArrayList<CreateBuildComponentDto>

    init {
        this.name = build.name
        this.description = build.description
        this.isPublic = true

        this.components = ArrayList()
        for ((_, buildComponents) in build.components) {
            for (buildComponent in buildComponents) {
                this.components.add(CreateBuildComponentDto(buildComponent.component.id, buildComponent.count))
            }
        }
    }
}